ouroboros-network/test/Test/Ouroboros/Network/PeerSelection.hs

```diff
@@ -752,6 +757,169 @@ check_governor_connstatus _ trace0 =
 -- Progress properties
 --
 
+-- | Check that the governor can hit (but not overshoot) its target for the
+-- number of root peers. This has to be bounded by what is possible: we cannot
+-- always find enough peers, and when we can, some of them fail.
+--
+prop_governor_target_root_below :: GovernorMockEnvironment -> Property
+prop_governor_target_root_below env =
+    let events = Signal.eventsFromListUpToTime (Time (10 * 60 * 60))
+               . selectPeerSelectionTraceEvents
+               . runGovernorInMockEnvironment
+               $ env
+
+        envTargetsSig :: Signal Int
+        envTargetsSig =
+          selectEnvTargets targetNumberOfRootPeers events
+
+        govLocalRootPeersSig :: Signal (Set PeerAddr)
+        govLocalRootPeersSig =
+          selectGovState (LocalRootPeers.keysSet . Governor.localRootPeers) events
+
+        govPublicRootPeersSig :: Signal (Set PeerAddr)
+        govPublicRootPeersSig =
+          selectGovState Governor.publicRootPeers events
+
+        govRootPeersSig :: Signal (Set PeerAddr)
+        govRootPeersSig = Set.union <$> govLocalRootPeersSig <*> govPublicRootPeersSig
+
+        -- There are no opportunities if we're at or above target
+        --
+        requestOpportunity target public roots
+          | Set.size roots >= target
+          = Set.empty
+
+          | otherwise
+          = public Set.\\ roots
+
+        requestOpportunities :: Signal (Set PeerAddr)
+        requestOpportunities =
+          requestOpportunity
+            <$> envTargetsSig
+            <*> govPublicRootPeersSig
+            <*> govRootPeersSig
+
+        requestOpportunitiesIgnoredTooLong :: Signal (Set PeerAddr)
+        requestOpportunitiesIgnoredTooLong =
+          Signal.keyedTimeout
+            10 -- seconds
+            id
+            requestOpportunities
+
+     in counterexample
+          ("\nSignal key: (target, local peers, public peers, root peers, " ++
+           "opportunities, ignored too long)") $
+
+        signalProperty 20 show
+          (\(_,_,_,_,_,toolong) -> Set.null toolong)
+          ((,,,,,) <$> envTargetsSig
+                   <*> govLocalRootPeersSig
+                   <*> govPublicRootPeersSig
+                   <*> govRootPeersSig
+                   <*> requestOpportunities
+                   <*> requestOpportunitiesIgnoredTooLong)
+
+-- | A variant of 'prop_governor_target_established_below' but for the target
+-- that any public root peers should become established.
+--
+-- We do not need separate above and below variants of this property since it
+-- is not possible to exceed the target.
+--
+prop_governor_target_established_public :: GovernorMockEnvironment -> Property
```

Comment:
Hmm, what if we adopt the following naming convention:
`prop_diffusion_target_established_below` where we use `prop_diffusion` instead of `prop_governor` with the same suffix as the original test.

Reply:
I am not sure why because these are not run in Diffusion, so it makes sense to be `prop_governor`

Reply:
you're right, ok.